spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/comet/CometIcebergColumnarBatchReader.java

```diff
@@ -0,0 +1,303 @@
+    this.readers =
+        readers.stream()
+            .map(CometIcebergColumnReader.class::cast)
+            .toArray(CometIcebergColumnReader[]::new);
+    this.hasIsDeletedColumn =
+        readers.stream().anyMatch(reader -> reader instanceof CometIcebergDeleteColumnReader);
+
+    AbstractColumnReader[] abstractColumnReaders = new AbstractColumnReader[readers.size()];
+    delegate = new BatchReader(abstractColumnReaders);
+    delegate.setSparkSchema(SparkSchemaUtil.convert(schema));
+  }
+
+  @Override
+  public void setRowGroupInfo(
+      PageReadStore pageStore, Map<ColumnPath, ColumnChunkMetaData> metaData, long rowPosition) {
+    for (int i = 0; i < readers.length; i++) {
+      try {
+        if (!(readers[i] instanceof CometIcebergConstantColumnReader)
+            && !(readers[i] instanceof CometIcebergPositionColumnReader)
+            && !(readers[i] instanceof CometIcebergDeleteColumnReader)) {
+          readers[i].reset();
+          readers[i].setPageReader(
+              pageStore.getPageReader(((CometIcebergColumnReader) readers[i]).getDescriptor()));
+        }
+      } catch (IOException e) {
+        throw new UncheckedIOException("Failed to setRowGroupInfo for Comet vectorization", e);
+      }
+    }
+
+    for (int i = 0; i < readers.length; i++) {
+      delegate.getColumnReaders()[i] = ((CometIcebergColumnReader) this.readers[i]).getDelegate();
+    }
+
+    this.rowStartPosInBatch = rowPosition;
+  }
+
+  public void setDeleteFilter(DeleteFilter<InternalRow> deleteFilter) {
+    this.deletes = deleteFilter;
+  }
+
+  @Override
+  public final ColumnarBatch read(ColumnarBatch reuse, int numRowsToRead) {
+    ColumnarBatch columnarBatch = new ColumnBatchLoader(numRowsToRead).loadDataToColumnBatch();
+    rowStartPosInBatch += numRowsToRead;
+    return columnarBatch;
+  }
+
+  @Override
+  public void setBatchSize(int batchSize) {
+    for (CometIcebergColumnReader reader : readers) {
+      if (reader != null) {
+        reader.setBatchSize(batchSize);
+      }
+    }
+  }
+
+  @Override
+  public void close() {
+    for (CometIcebergColumnReader reader : readers) {
+      if (reader != null) {
+        reader.close();
+      }
+    }
+  }
+
+  private class ColumnBatchLoader {
+    private final int numRowsToRead;
+    // the rowId mapping to skip deleted rows for all column vectors inside a batch, it is null when
+    // there is no deletes
+    private int[] rowIdMapping;
+    // the array to indicate if a row is deleted or not, it is null when there is no "_deleted"
+    // metadata column
+    private boolean[] isDeleted;
+
+    ColumnBatchLoader(int numRowsToRead) {
+      Preconditions.checkArgument(
+          numRowsToRead > 0, "Invalid number of rows to read: %s", numRowsToRead);
+      this.numRowsToRead = numRowsToRead;
+      if (hasIsDeletedColumn) {
+        isDeleted = new boolean[numRowsToRead];
+      }
+    }
+
+    ColumnarBatch loadDataToColumnBatch() {
```

Comment:
How much of this overlaps with `ColumnarBatchReader`?

Reply:
There are lots of overlap. I will put the common code in a base method and reuse.